File: app/src/main/java/com/jbrunton/mymovies/app/moviedetails/MovieDetailsViewStateFactory.java
package com.jbrunton.mymovies.app.moviedetails;

import com.jbrunton.mymovies.app.movies.BaseMovieViewStateFactory;
import com.jbrunton.mymovies.app.movies.MovieViewState;
import com.jbrunton.mymovies.app.shared.LoadingViewState;
import com.jbrunton.mymovies.models.Movie;

public class MovieDetailsViewStateFactory extends BaseMovieViewStateFactory {
    public MovieDetailsViewState fromMovie(Movie movie) {
        return MovieDetailsViewState.builder()
                .setLoadingViewState(LoadingViewState.OK_STATE)
                .setMovie(toMovieViewState(movie))
                .build();
    }

    public MovieDetailsViewState fromError(Throwable throwable) {
        return MovieDetailsViewState.builder()
                .setLoadingViewState(loadingViewStateFactory.fromError(throwable))
                .setMovie(MovieViewState.EMPTY)
                .build();
    }

    public MovieDetailsViewState loadingState() {
        return MovieDetailsViewState.builder()
                .setLoadingViewState(LoadingViewState.LOADING_STATE)
                .setMovie(MovieViewState.EMPTY)
                .build();
    }

    private MovieViewState toMovieViewState(Movie movie) {
        return setDefaults(MovieViewState.builder(), movie)
                .overview(movie.overview().get())
                .build();
    }
}

File: app/src/main/java/com/jbrunton/mymovies/moviedetails/MovieDetailsViewStateFactory.kt
package com.jbrunton.mymovies.moviedetails

import com.jbrunton.entities.Movie
import com.jbrunton.mymovies.movies.MovieViewState
import com.jbrunton.mymovies.movies.from
import com.jbrunton.mymovies.shared.Loading
import com.jbrunton.mymovies.shared.LoadingViewState
import com.jbrunton.mymovies.shared.LoadingViewStateFactory
import com.jbrunton.mymovies.shared.Success

class MovieDetailsViewStateFactory {
    private val loadingViewStateFactory = LoadingViewStateFactory()

    fun fromMovie(movie: Movie) = Success(toMovieViewState(movie))

    fun fromError(throwable: Throwable): LoadingViewState<MovieViewState> = LoadingViewState.fromError(throwable)

    fun loadingState(): LoadingViewState<MovieViewState> = Loading()

    private fun toMovieViewState(movie: Movie): MovieViewState {
        return MovieViewState.Builder().from(movie).apply {
            this.overview = movie.overview.get()
        }.build()
    }
}
